### PR TITLE
Add stack to MiddlewareDispatcher

### PR DESCRIPTION
**Pull request by yiiliveext**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Tests pass?   | ✔️

You have to change `MiddlewareDispatcherFactory` in `yii-demo` after this PR.
 
```php
class MiddlewareDispatcherFactory
{
    public function __invoke(ContainerInterface $container)
    {
        $session = $container->get(SessionMiddleware::class);
        $router = $container->get(Router::class);
        $errorCatcher = $container->get(ErrorCatcher::class);
        $subFolder = $container->get(SubFolder::class);
        $dispatcher = new MiddlewareDispatcher($container);

        return $dispatcher
            ->addMiddleware($router)
            ->addMiddleware($subFolder)
            ->addMiddleware($session)
            ->addMiddleware($errorCatcher);
    }
}
```

### CODE REVIEW

**Review by samdark — APPROVED** [review]

**Comment by yiiliveext** [comment]
I think we could make the process() method protected.

File: src/MiddlewareDispatcher.php [hunk 1]
```diff
@@ -41,60 +41,70 @@ public function __construct(
 
         $this->container = $container;
 
-        foreach ($middlewares as $middleware) {
-            $this->add($middleware);
+        for ($i = count($middlewares) - 1; $i >= 0; $i--) {
```
Comment (viktorprogger) [inline src/MiddlewareDispatcher.php]: Why did you change `foreach` to `for`?

Comment (viktorprogger) [inline src/MiddlewareDispatcher.php]: Why did you change `foreach` to `for`?

Comment (yiiliveext) [reply src/MiddlewareDispatcher.php]: Hmm... I think this is faster then foreach+array_reverse().

**Comment by roxblnfk** [comment]
I suggest:

`MiddlewareDispatcher` rename to `Pipeline`

For the `process()` method add `MiddlewareInterface`
The `process()` method should throw an exception if the stack of middlewares is empty

The `dispatch()` method rename to `handle()` and add `RequestHandlerInterface`
The `handle()` method should throw an exception if the $nextHandler is not defined

The `addMiddleware()` rename to `add()`
Remove `$middlewares` argument from the `__construct()` method

File: src/MiddlewareDispatcher.php [hunk 2]
```diff
@@ -13,88 +13,89 @@
...
-            $this->middlewares[] = $middleware;
-        } else {
+            $middleware = new Callback($middleware, $this->container);
+        }
+
+        if (!$middleware instanceof MiddlewareInterface) {
             throw new \InvalidArgumentException('Middleware should be either callable or MiddlewareInterface instance. ' . get_class($middleware) . ' given.');
         }
+
+        array_unshift($this->middlewares, $middleware);
```
Comment (roxblnfk) [inline src/MiddlewareDispatcher.php]:
```suggestion
        $this->middlewares[] = $middleware;
```

File: src/MiddlewareDispatcher.php [hunk 3]
```diff
@@ -13,88 +13,89 @@
...
-    public function handle(ServerRequestInterface $request): ResponseInterface
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
     {
-        $middleware = current($this->middlewares);
-        next($this->middlewares);
-        if ($middleware === false) {
-            if ($this->nextHandler !== null) {
-                return $this->nextHandler->handle($request);
+        if ($this->stack === null) {
+            for ($i = count($this->middlewares) - 1; $i >= 0; $i--) {
```
Comment (roxblnfk) [inline src/MiddlewareDispatcher.php]:
```suggestion
            foreach ($this->middlewares as $middleware) {
```

File: src/MiddlewareDispatcher.php [hunk 4]
```diff
@@ -13,88 +13,89 @@
...
+                $handler = $this->wrap($this->middlewares[$i], $handler);
```
Comment (roxblnfk) [inline src/MiddlewareDispatcher.php]:
```suggestion
                $handler = $this->wrap($middleware, $handler);
```

**Review by roxblnfk — CHANGES_REQUESTED** [review]

File: src/MiddlewareDispatcher.php (hunk 1 above)
Comment (viktorprogger) [reply src/MiddlewareDispatcher.php]: 1. You don't need `array_reverse` here
1. [Foreach is faster a bit](https://github.com/viktorprogger/foreach-bench) (without `array_reverse`) and more readable: 
image

**Review by viktorprogger — CHANGES_REQUESTED** [review]

File: src/MiddlewareDispatcher.php (hunk 3 above)
Comment (yiiliveext) [reply src/MiddlewareDispatcher.php]: I thought exactly the same but I so lazy...

File: src/MiddlewareDispatcher.php (hunk 2 above)
Comment (yiiliveext) [reply src/MiddlewareDispatcher.php]: I thought exactly the same but I so lazy...

File: src/MiddlewareDispatcher.php (hunk 4 above)
Comment (yiiliveext) [reply src/MiddlewareDispatcher.php]: Done.

**Comment by samdark** [comment]
@roxblnfk nice ideas above but let's handle these separately, OK?

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thanks!
